### PR TITLE
Add RSNv3 for CA

### PR DESCRIPTION
**Pull request by edewata**
The `CertificateRepository`, `RequestRepository`, and `Repository` classes have been modified to support the new RSNv3 and the legacy ID generators. `pkispawn`'s `default.cfg` has been modified to provide the configuration parameters.

Results:
https://github.com/dogtagpki/pki/runs/5296272839?check_suite_focus=true#step:13:9
https://github.com/dogtagpki/pki/runs/5296272839?check_suite_focus=true#step:14:9

Doc:
https://github.com/dogtagpki/pki/wiki/Installing-CA-with-Random-Serial-Numbers-v3

### CODE REVIEW

**Comment by rcritten** [comment]
So for the IPA type encoding issue, does this mean we can set pki_cert_id_length to a value that "fits" into IPA even though it would restrict the total number of available certificates?

**Comment by edewata** [comment]
@rcritten Yes, but @tiran has said before that the minimum should be 65:
https://pagure.io/freeipa/issue/8802#comment-727012

**Comment by rcritten** [comment]
> @rcritten Yes, but @tiran has said before that the minimum should be 65: https://pagure.io/freeipa/issue/8802#comment-727012

I read his comment more about the minimum for a safe random distribution without collisions.

**Comment by edewata** [comment]
@rcritten Yes, for testing feel free to use any length, but for production I'd suggest to use at least 65 as @tiran suggested to avoid collisions.

**Review by ckelleyRH — COMMENTED** [review]
Seems to work well, tested on my machine also. I have a few questions about the implementation but nothing major.

You said `pkispawn` had been modified in the description but it has not, did you mean the configuration in general?

File: base/server/src/main/java/com/netscape/cmscore/dbs/CertificateRepository.java [hunk 1]
```diff
@@ -82,6 +82,9 @@
     private static final String PROP_MINIMUM_RANDOM_BITS = "minimumRandomBits";
     private static final BigInteger BI_MINUS_ONE = (BigInteger.ZERO).subtract(BigInteger.ONE);
```
[the remark this answers is not in the document]
Comment (edewata) [reply base/server/src/main/java/com/netscape/cmscore/dbs/CertificateRepository.java]: OK, updated.

File: base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java [hunk 2]
```diff
@@ -166,11 +175,27 @@ protected void setLastSerialNo(BigInteger lastSN) {
         mLastSerialNo = lastSN;
     }
 
+    public void setIDGenerator(String generator) throws Exception {
```
[the remark this answers is not in the document]
Comment (edewata) [reply base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java]: Updated.

File: base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java [hunk 3]
```diff
@@ -54,6 +55,9 @@
 
     public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(Repository.class);
 
+    public static final int LEGACY = 0;
+    public static final int RANDOM = 1;
```
[the remark this answers is not in the document]
Comment (edewata) [reply base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java:59]: There's always a possibility, but we may never know. The legacy code was using a boolean variable to switch between sequential and random  (i.e. before it's called RSNv1), and it turned out to be insufficient for RSNv3:

* https://github.com/dogtagpki/pki/blob/master/base/ca/shared/conf/CS.cfg#L769
* https://github.com/dogtagpki/pki/blob/master/base/server/src/main/java/com/netscape/cmscore/dbs/CertificateRepository.java#L87

File: base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java [hunk 4]
```diff
@@ -268,6 +293,21 @@ public void setTheSerialNumber(BigInteger num) throws EBaseException {
...
+
+            logger.debug("Repository: Generating random serial number");
+
+            BigInteger id = new BigInteger(idLength, secureRandom);
+            logger.info("Repository: id: 0x" + id.toString(16));
+
+            // JSS BigInt does not allow negative value
+            // so use the absolute (i.e. positive) value.
+            id = id.abs();
+            logger.info("Repository: id: 0x" + id.toString(16));
```
[the remark this answers is not in the document]
Comment (edewata) [reply base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java]: Yeah, I forgot to remove it after testing. Updated.

**Comment by edewata** [comment]
@ckelleyRH Thanks for the comments. Please see the updated PR.

I changed the commit description to say "`pkispawn`'s `default.cfg` has been modified ..." for clarity because most people probably do not use the `default.cfg` directly, but they will see that `pkispawn` provides some new params.

File: base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ckelleyRH) [reply base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java:59]: Yeah I saw that one. I don't think I understand why the Boolean is insufficient but it doesn't matter that much. The code works and leaving the door open for extension is not a crime :smile:

**Review by ckelleyRH — APPROVED** [review]
@edewata Thanks for updating - I think it looks good now!

File: base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (edewata) [reply base/server/src/main/java/com/netscape/cmscore/dbs/Repository.java:59]: The `mEnableRandomSerialNumbers` can only support two modes, but we need three: sequential, RSNv1, and RSNv3. We also did discuss RSNv2 in the past but decided not to implement it. So there could be multiple options, but I don't think we want to keep adding a separate `Boolean` var for each option.

**Comment by edewata** [comment]
@ckelleyRH Thanks!

**Pull request merged**